libraries/standard/mqtt/src/mqtt.c

```diff
@@ -948,14 +954,33 @@ static MQTTStatus_t receiveConnack( MQTTContext_t * const pContext,
                                                       pContext->transportInterface.networkContext,
                                                       pIncomingPacket );
 
-        /* Loop until there is data to read or if the timeout has not expired. */
-    } while( ( status == MQTTNoDataAvailable ) &&
-             ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) < timeoutMs ) );
+        /* The loop times out based on 2 conditions.
+         * 1. If a non-NULL getTime function is provided by the application:
+         *    Loop times out based on the timeout calculated by getTime()
+         *    function.
+         * 2. If a non-NULL getTime function is not provided by the application:
+         *    Loop times out based on the maximum number of retries config
+         *    MQTT_CONNACK_RECEIVE_RETRY_COUNT. */
+        if( getTimeStamp != NULL )
+        {
+            /* Loop runs only once for 0 timeoutMs. */
+            breakFromLoop = ( bool ) ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs );
+        }
+        else
+        {
+            breakFromLoop = ( bool ) ( loopCount++ >= MQTT_CONNACK_RECEIVE_RETRY_COUNT );
+        }
+
+        /* Loop until there is data to read or if we have exceeded the timeout/retries. */
+    } while( ( status == MQTTNoDataAvailable ) && ( !breakFromLoop ) );
 
     if( status == MQTTSuccess )
     {
         /* Time taken in this function so far. */
-        timeTakenMs = calculateElapsedTime( getTimeStamp(), entryTimeMs );
+        if( getTimeStamp != NULL )
```

Comment:
In the event of `getTimeStamp` being `NULL`, then `timeTakenMs` won't be updated and be 0, the condition `timeTakenMs < timeoutMs` will hold, and `remainingTimeMs` will be set to `timeoutMs`. Then `receivePacket` will be called with a nonzero remaining time, and will have the potential to be an infinite loop if no data can be received in `recvExact`

Reply:
I think you are mentioning about a case in which timeoutMs is non zero and getTime is NULL. Ideally the timeoutMs doesn't matter in that case, but as you have pointed out our logic may not be handling it. I will zero out timeoutMs if getTime is NULL